primitives/npos-elections/src/pjr.rs

```diff
@@ -0,0 +1,607 @@
+fn pre_score<AccountId: IdentifierT>(
+	unelected: CandidatePtr<AccountId>,
+	voters: &[Voter<AccountId>],
+	t: Threshold,
+) -> ExtendedBalance {
+	debug_assert!(!unelected.borrow().elected);
+	voters
+		.iter()
+		.filter(|ref v| v.votes_for(&unelected.borrow().who))
+		.fold(Zero::zero(), |acc: ExtendedBalance, voter| acc.saturating_add(slack(voter, t)))
+}
+
+
+/// The slack of a voter at a given state.
+///
+/// The slack of each voter, with threshold `t` is the total amount of stake that this voter can
+/// spare to a new potential member, whilst not dropping the backing stake of any of its currently
+/// active members below `t`. In essence, for each of the current active candidates `c`, we assume
+/// that we reduce the edge weight of `voter` to `c` from `w` to `w * min(1 / (t / support(c)))`.
+///
+/// More accurately:
+///
+/// 1. If `c` exactly has `t` backing or less, then we don't generate any slack.
+/// 2. If `c` has more than `t`, then we reduce it to `t`.
+fn slack<AccountId: IdentifierT>(voter: &Voter<AccountId>, t: Threshold) -> ExtendedBalance {
+	let budget = voter.budget;
+	let leftover = voter.edges.iter().fold(Zero::zero(), |acc: ExtendedBalance, edge| {
+		let candidate = edge.candidate.borrow();
+		if candidate.elected {
+			// TODO: using perbill here is just going to cause annoyance, why not just subtract?
+			let extra =
+				Perbill::one().min(Perbill::from_rational_approximation(t, candidate.backed_stake))
+				* edge.weight;
+			acc.saturating_add(extra)
+		} else {
+			// No slack generated here.
+			acc
+		}
+	});
+
+	// NOTE: candidate for saturating_log_sub()
+	budget.saturating_sub(leftover)
+}
+
+/// Compute the threshold corresponding to the standard PJR property
+///
+/// `t-PJR` checks can check PJR according to an arbitrary threshold. The threshold can be any value,
+/// but the property gets stronger as the threshold gets smaller. The strongest possible `t-PJR` property
+/// corresponds to `t == 0`.
+///
+/// However, standard PJR is less stringent than that. This function returns the threshold whose
+/// strength corresponds to the standard PJR property.
+///
+/// - `committed_size` is the number of winners of the election.
+/// - `weights` is an iterator of voter stakes. If the sum of stakes is already known,
+///   `std::iter::once(sum_of_stakes)` is appropriate here.
+pub fn standard_threshold(
+	committeed_size: usize,
+	weights: impl IntoIterator<Item = ExtendedBalance>,
+) -> Threshold {
+	weights
+		.into_iter()
+		.sum::<Threshold>()
+	/ committeed_size as Threshold
+}
+
+/// Check a solution to be PJR.
+///
+/// The PJR property is true if `t-PJR` is true when `t == sum(stake) / committee_size`.
+///
+/// This function returns an `AccountId` in the `Err` case. This is the counterexample: the ID of the
+/// unelected candidate with the highest prescore, such that `pre_score(counterexample) >= t`.
+pub fn pjr_check<AccountId: IdentifierT>(
```

Comment:
nit: I think the order of functions in this file is not logical. I'd say it should be high -> low abstraction. So Iexpect to first see the functions that auto-compute the `t`. Then, the functions that accept `t`. Then, `_core` stuff, and finally `slack` and `prescore`. 

Would be nice if you fix this in the other PR as well.

Reply:
b36f71895 and c1f1ca718